Ble privacy scanning initiating

pan- opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

Follow up to #13560 .

Addition of private addresses management when initiating connection and when scanning. 

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

pan- committed:
BLE: Add callback for scan started and stoped.

pan- committed:
BLE: Update state when scan started/stopped callback received.

pan- committed:
BLE: Fix random static address storage

pan- committed:
BLE: check LL resolution availability before enabling it.

pan- committed:
BLE: Add routine that computes the random address to use depending on the procedure.

pan- committed:
BLE: Implement private address rotation for initiating.

pan- committed:
BLE: Private address when initiating connection

pan- committed:
BLE: privacy tracing cleanup

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/cordio/source/PalGapImpl.cpp:
```diff
@@ -683,16 +683,34 @@ void PalGap::gap_handler(const wsfMsgHdr_t *msg)
+            if (!handler) { break; }
+
+            if (get_gap().ext_scan_stopping) {
+                get_gap().ext_scan_stopping = false;
```
this should be set even if there is not handler, so the handler check above needs to be done later

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -461,23 +463,21 @@ ble_error_t Gap::stopScan()
     ble_error_t err;
 
-    if (!_scan_enabled) {
+    if ((!_scan_enabled && !_scan_pending) || _scan_pending) {
```
so what happens when scan is pending and we cancel it and it returns error none but it actually didn't cancel anything?

pan- replied:
If scan is pending then it can't be canceled by the user. The callback `on_scan_stopped` always reset the `_scan_pending` flag whether the stop operation as completed or not.  
The `_scan_enabled` flag, it is reset when `on_scan_stopped` is called regardless of the operation success. I though of putting an error as we can't retrieve the scanning status from the controller but ended up doing the same choice as the cordio stack. 
What do you think ?

paul-szczepanek-arm replied:
IMHO: if scan pending return invalid state

paul-szczepanek-arm replied:
or some other error to let user know this call is not available at this time

pan- replied:
I will return `BLE_STACK_BUSY`

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1008,37 +1031,85 @@ Gap::GapShutdownCallbackChain_t &Gap::onShutdown()
+        _scan_enabled = true;
+    } else {
+        _scan_pending = false;
+        _scan_enabled = true;
```
```suggestion
        _scan_enabled = false;
```

paul-szczepanek-arm replied:
or simpler
```
_scan_pending = false;
_scan_enabled = success;
```
without the if

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1008,37 +1031,85 @@ Gap::GapShutdownCallbackChain_t &Gap::onShutdown()
+void Gap::on_scan_stopped(bool success)
+{
+    if (!success) {
+        _scan_pending = false;
```
if you put the if statement lower then you don't have to duplicate it

paul-szczepanek-arm reviewed (changes requested):
I need to review this again, the states are really complex

pan- committed:
BLE: return stack busy instead of invalid state when applicable

enabling/disabling scanning and advertising operations can be prevented if the previous operations hasn't completed.
Return that the stack is busy and cannot fulfill the request at the moment rather than the state is invalid.

pan- committed:
BLE: Fix state update when event handler not present

pan- committed:
BLE: improve scan state bookkeeping
